decisions/0001-bundle-granularity.rst

```diff
@@ -0,0 +1,199 @@
+course may potentially have thousands of directories.
+
+The major focus of this approach is operational simplicity at scale. It is
+generally less featureful than other proposals and shifts more of the work to
+OLX-aware clients.
+
+An important mental shift is that this walks back from some of the file system
+analogies of the original proposal. We still have files in Bundles, and XBlock
+content is still stored in directories with their associated static assets. But
+when we make data references across XBlocks, it's by identifiers and conventions
+and not by relative file paths. It's the job of clients to transform that into
+a more friendly file-path based format for the human readable format.
+
+XBlock Directories
+==================
+
+XBlock Directories are a client convention and not a core Blockstore primitive.
+Each directory represents a single XBlock and its associated assets. The naming
+scheme is::
+
+    /{definition-key}/
+                     {tag-type}.xml
+                     static/
+
+The ``static`` directory is optional. For a concrete example::
+
+    /problem+blockstore_granularity/
+                                    problem.xml
+                                    static/
+                                           diagram.png
+
+Meanwhile, a container looks like:
+
+    /unit+blockstore_big_questions/problem.xml
+
+When containers reference their children, it's done via a new tag. So a
+particular ``unit.xml`` might look like::
+
+    <unit display_name="">
+        <!-- Normal usage -->
+        <xblock-include definition="unit+blockstore_big_questions" />
+
+        <!-- If we want to include the same content with a modifier to the usage
+             key so that we can include it multiple times with different state
+             storage. -->
+        <xblock-include definition="unit+blockstore_big_questions" usage="alternate" />
+
+        <!-- XBlock that exists in a Linked Bundle aliased to "arch_hangout_vidoes" -->
+        <xblock-include definition="arch_hangout_videos/video+blockstore_overview"/>
+    </unit>
+
+Links
+=====
+
+Link information is stored at the Snapshot level. Since the human-readable
+format is going to be a transform, we don't have to make symlink analogies to
+make the data easier to work with in authoring. Instead, Links are a mapping of
+names to Bundle Snapshots::
+
+    {
+        "links": {
+            "arch_hangout_videos": {
+                "bundle_uuid": "3fcf5f61-bc23-41ec-9452-26d12dc3b13c",
+                "snapshot": "8f7bc6e89581591fa925fefa3819d382fd793839"
+                "dependencies": [
+                    {
+                        "bundle_uuid": "fd7ee4b3-0540-406c-98b9-dd050b7ddcb2",
+                        "snapshot": "c0c0940e4b3151908b60cecd1ef5e2aa19904676"
+                    }
+                ]
+            }
+        }
+    }
+
+Since all Snapshots capture their full dependencies, they can do cycle detection
```

Comment:
Would there be cycle detection between XBlocks in the Bundle?

Reply:
I think it should be fairly straightforward for the client to do so?